src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFileImpl.java

```diff
@@ -0,0 +1,133 @@
+package edu.kpi.testcourse.storage;
+
+import com.google.gson.reflect.TypeToken;
+import edu.kpi.testcourse.entities.UrlAlias;
+import edu.kpi.testcourse.logic.UrlShortenerConfig;
+import edu.kpi.testcourse.serialization.JsonTool;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * A file-backed implementation of {@link UrlRepository} suitable for use in production.
+ */
+public class UrlRepositoryFileImpl implements UrlRepository {
+
+  // Urls, keyed by alias.
+  private final Map<String, UrlAlias> urlMapByAlias;
+  // All User urls, keyed by email
+  private final Map<String, List<UrlAlias>> urlsMapByEmail;
```

Comment:
Could you explain what the reason of using one more HashMap keyed by email instead of iterating another one is?

Reply:
Because it's two different HashMaps which get different values.

Reply:
But from the perspective of consumed memory, this approach loses significantly

Reply:
But by adding one more field we are winning in cpu time by saving one extra "for" loop (if you only knew how much memory java consumes while launching... Memory which is consumed by adding yet another field is just nothing.)

Reply:
Just the point is, what your field contains I guess. When searching, indexing etc. stores the elements only as value and maintain internally the indexing for every element. While HashMap stores elements with key and value pair, i.e. two objects. So HashMap takes more memory comparatively.

Reply:
We store the same objects in these hashmaps, so the value of the hashmap is a link to the object, not the object itself. The memory usage is low in this case. Moreover, search by list takes an O(n) time, while hashmap search time is O(1). Also in real applications, read operations are the most common, so it makes sense to complicate writing and simplify reading. So having an extra hashmap is a better choice.